test/src/test/java/org/corfudb/runtime/BackupRestoreIT.java

```diff
@@ -0,0 +1,338 @@
+            assertThat(aValueSet.get(i)).isEqualTo(bValueSet.get(i));
+        }
+    }
+
+    /**
+     * Implement an end-to-end Backup and Restore test for multiple tables.
+     *
+     * 1. Open multiple tables and generate random entries.
+     * 2. Backup a list of tables and obtain a tar file.
+     * 3. Use the tar file to restore tables.
+     * 4. Compare the table contents before and after the backup/restore.
+     */
+    @Test
+    public void backupRestoreMultipleTablesTest() throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+
+        // Set up the test environment
+        setupEnv();
+
+        // Create Corfu Store to add entries into server
+        CorfuStore srcDataCorfuStore = new CorfuStore(srcDataRuntime);
+        CorfuStore destDataCorfuStore = new CorfuStore(destDataRuntime);
+
+        List<String> tableNames = getTableNames(numTables);
+
+        // Generate random entries and save into sourceServer
+        for (String tableName : tableNames) {
+            generateData(srcDataCorfuStore, tableName, false);
+        }
+
+        // Obtain the corresponding streamIDs for the tables in sourceServer
+        List<UUID> streamIDs = new ArrayList<>();
+        for (String tableName : tableNames) {
+            streamIDs.add(CorfuRuntime.getStreamID(TableRegistry.getFullyQualifiedTableName(NAMESPACE, tableName)));
+        }
+
+        // Backup
+        Backup backup = new Backup(BACKUP_TAR_FILE_PATH, streamIDs, backupRuntime);
+        backup.start();
+
+        // Verify that backup tar file exists
+        File backupTarFile = new File(BACKUP_TAR_FILE_PATH);
+        assertThat(backupTarFile).exists();
+
+        // Restore using backup files
+        Restore restore = new Restore(BACKUP_TAR_FILE_PATH, restoreRuntime);
+        restore.start();
+
+        // Compare data entries in CorfuStore before and after the Backup/Restore
+        for (String tableName : tableNames) {
+            openTableWithoutBackupTag(destDataCorfuStore, tableName);
+            compareCorfuStoreTables(srcDataCorfuStore, tableName, destDataCorfuStore, tableName);
+        }
+
+        // Close servers and runtime before exiting
+        cleanEnv();
+    }
+
+    /**
+     * Backup and restore tables which have requires_backup_support tag set
+     */
+    @Test
+    public void backupRestoreTaggedTables() throws IOException, NoSuchMethodException, IllegalAccessException, InvocationTargetException {
+        // Set up the test environment
+        setupEnv();
+
+        // Create Corfu Store to add entries into server
+        CorfuStore srcDataCorfuStore = new CorfuStore(srcDataRuntime);
+        CorfuStore destDataCorfuStore = new CorfuStore(destDataRuntime);
+
+        List<String> tableNames = getTableNames(numTables);
+
+        // Generate random entries and save into sourceServer
+        for (String tableName : tableNames) {
+            generateData(srcDataCorfuStore, tableName, true);
```

Comment:
we only have a test case where all tables are tagged as required for backup, can we have a test that combines non-tagged with tagged? so we validate that it is selective.

Reply:
Done. Thanks.